presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java

```diff
@@ -0,0 +1,231 @@
+    }
+
+    @Override
+    public long getLong(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(BIGINT, INTEGER));
+        return Long.valueOf(String.valueOf(getFieldValue(field)));
+    }
+
+    @Override
+    public double getDouble(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(DOUBLE));
+        return (Double) getFieldValue(field);
+    }
+
+    @Override
+    public Slice getSlice(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(VARCHAR));
+
+        Object value = getFieldValue(field);
+        ElasticsearchColumnHandle column = columnHandles.get(field);
+        if (tupleDomain.getDomains().isPresent()) {
+            Map<ColumnHandle, Domain> domains = tupleDomain.getDomains().get();
+            Domain domain = domains.get(column);
+            if (domain != null) {
+                for (Range range : domain.getValues().getRanges().getOrderedRanges()) {
+                    if (range.isSingleValue()) {
+                        return utf8Slice(String.valueOf(getValue(column.getColumnType(), range.getLow().getValue())));
+                    }
+                }
+            }
+        }
+        if (value instanceof Collection) {
+            return utf8Slice(new JSONArray((List<Map<String, Object>>) value).toString());
+        }
+        return utf8Slice(String.valueOf(value));
+    }
+
+    @Override
+    public Object getObject(int field)
+    {
+        return null;
+    }
+
+    @Override
+    public boolean isNull(int field)
+    {
+        checkArgument(field < columnHandles.size(), "Invalid field index");
+        return getFieldValue(field) == null;
+    }
+
+    private void checkFieldType(int field, Set<Type> expectedTypes)
+    {
+        Type actual = getType(field);
+        checkArgument(expectedTypes.contains(actual), "Field %s expected type inconsistent with %s", field, actual);
+        StringBuilder builder = new StringBuilder();
+        for (Type type : expectedTypes) {
+            builder.append(", ");
+            builder.append(type.getDisplayName());
+        }
+    }
+
+    @Override
+    public void close()
+    {
+    }
+
+    private List<SearchHit> sendElasticQuery(ElasticsearchQueryBuilder queryBuilder)
+    {
+        List<SearchHit> result = new ArrayList<>();
+        SearchResponse response = queryBuilder.buildScrollSearchRequest().execute().actionGet(timeout.toMillis());
+
+        if (response.getHits().getTotalHits() > maxHits) {
+            throw new PrestoException(ELASTIC_SEARCH_EXCEEDS_MAX_HIT_ERROR, "The number of hits for the query: " + response.getHits().getTotalHits() + " exceeds the configured max hits: " + maxHits);
+        }
+
+        while (true) {
+            Collections.addAll(result, response.getHits().getHits());
+            response = queryBuilder.prepareSearchScroll(response.getScrollId()).execute().actionGet(timeout.toMillis());
+            if (response.getHits().getHits().length == 0) {
```

Comment:
Can `response.getHits().getHits()` be `null`?  If yes, how should we handle that?

Reply:
I see that `response` has a `status()` method, so if one of the responses is "bad" (failed, timeout, etc.) should we continue sending requests as we do here or just stop? (I am not familiar with the ES API)